### PR TITLE
Fix a crash that happens when Firestore is being terminated when the last shared pointer to it is in a listener

### LINKED ISSUE #6909: Crash: NSInternalInconsistencyException', reason: 'FIRESTORE INTERNAL ASSERTION FAILED: Enqueue methods cannot be called when we are already running on target executor
### [REQUIRED] Step 1: Describe your environment

  * Xcode version: 12.1
  * Firebase SDK version: 7.0.0
  * Installation method: `Carthage`
  * Firebase Component: Auth, Core

#### Steps to reproduce:

Observing a crash when trying to delete a `FirebaseApp` object.

calling 
```
app.delete {
}
```

where app is a FirebaseApp

Seeing this
```
*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'FIRESTORE INTERNAL ASSERTION FAILED: Enqueue methods cannot be called when we are already running on target executor (this queue's executor: 'com.google.firebase.firestore.httpsinstacartfirebaseiocomv1', current executor: 'com.google.firebase.firestore.httpsinstacartfirebaseiocomv1') (expected !is_operation_in_progress_ || !executor_->IsCurrentExecutor())'
```

### PR DESCRIPTION
**Pull request by var-const**
If a user calls terminate and immediately disposes their reference to Firestore, it's possible that while termination is in process, the last remaining reference (more precisely, shared pointer) to Firestore is in a listener. When that listener is destroyed as part of the termination, it leads to `api::Firestore` being destroyed. Importantly, termination happens on the worker queue. The destructor of `api::Firestore` calls `Dispose` which presumes it's not called from the worker queue and tries to enqueue work, leading to a failing assertion and a crash.

The solution is simply to remove the sequential order checks when the queue enters/is in restricted mode. There is a legitimate case where the Firestore destructor can run on the worker queue, as the original issue shows. The complexity of #7412 also indicates that a simpler solution is preferable.

Fixes #6909.

### CODE REVIEW

**Comment by var-const** [comment]
@wilhuff I have reworked the code based on your feedback, and it seemed easier to start from scratch.

File: Firestore/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+# Unreleased
+- [fixed] Fixed a crash that would happen when the app is being deleted and
```
Comment (wilhuff) [inline Firestore/CHANGELOG.md:2]: Is there a way to simplify this phrasing? Both "the app is being deleted" and "immediately disposed of" sound duplicative. Also, should we capitalize "app"?

**Review by wilhuff — APPROVED** [review]
LGTM

**Comment by wilhuff** [comment]
The new test is failing:

> /Users/runner/work/firebase-ios-sdk/firebase-ios-sdk/build/../Firestore/Example/Tests/Util/XCTestCase+Await.mm:40: error: -[FIRDatabaseTests testDeleteAppWorksWhenLastReferenceToFirestoreIsInListener] : Asynchronous wait failed: Exceeded timeout of 25 seconds, with unfulfilled expectations: "setData".

**Comment by var-const** [comment]
> The new test is failing

Not sure why -- it was timing out on writing the doc to the backend, even though it's pretty much copied without (and succeeds when I run it locally). Anyway, it turns out that writing a doc isn't actually necessary to trigger the bug -- the fact that the client gets initialized when creating a document reference is enough -- so I removed the writing part, which fixed the CI issue. I verified that the shorter version of the test is enough to trigger the crash (without the fix from this PR).

File: Firestore/CHANGELOG.md (hunk 1 above)
Comment (var-const) [reply Firestore/CHANGELOG.md:2]: Sorry, I got caught up in CI failures and somehow missed this comment. Addressed in #7449.

**Pull request merged**